File: Java 8/BinarySearch.java
class BinarySearch
{
    // find out if a key x exists in the sorted array A using iterative binary search algorithm
    public static int binarySearch(int[] A, int x)
    {
        
        int left = 0, right = A.length - 1;

        
        while (left <= right)
        {
            // we find the mid value in the search space and compare it with key value

            int mid = (left + right) / 2;

            /* overflow can happen. Use:
             int mid = left + (right - left) / 2;
             int mid = right - (right - left) / 2; */

            // key value is found
            if (x == A[mid]) {
                return mid;
            }

            // discard all elements in the right search space including the mid element
            else if (x < A[mid]) {
                right = mid - 1;
            }

            // discard all elements in the left search space
            // including the mid element
            else {
                left = mid + 1;
            }
        }

        // x doesn't exist in the array
        return -1;
    }

    public static void main(String[] args)
    {
        int[] A = {1, 4, 9, 16, 25, 36};
        int key = 4;

        int index = binarySearch(A, key);

        if (index != -1) 
	{
		System.out.println("Element found at index " + index);
	} 
	
	else 
	{
		System.out.println("Element not found in the array");
	}
    }
}

File: CPP/binary_search.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	int n;
	cin >> n;
	int ar[n];
	for(int i=0;i<n;++i) cin >> ar[i];
	sort(ar, ar + n);

	int target;
	cin >> target;
	int hi = n-1, lo = 0;
	while(lo <= hi) {
		int mid = (hi+lo)/2;
		if(ar[mid] == target) {
			cout << "Found!\n";
			return 0;
		}
		if(target > ar[mid]) lo = mid + 1;
		else hi = mid - 1;
	}
	cout << "Not Found!\n";
	return 0;
}